versioning/BRM/brmshmimpl.cpp

```diff
@@ -232,7 +240,166 @@ void BRMShmImpl::destroy()
     if (!oldName.empty()) bi::shared_memory_object::remove(oldName.c_str());
 }
 
+BRMManagedShmImpl::BRMManagedShmImpl(unsigned key, off_t size, bool readOnly)
+    : BRMShmImplParent(key, size, readOnly)
+{
+    string keyName = ShmKeys::keyToName(fKey);
+    off_t curSize = 0;
+
+    for(uint32_t tries = 0; fSize == 0 && tries <= ShmCreateMaxRetries; ++tries)
+    {
+        try
+        {
+            auto* shmSegment = new boost::interprocess::managed_shared_memory(bi::open_only, keyName.c_str());
+            curSize = shmSegment->get_size();
+
+            if (curSize == 0)
+            {
+                delete shmSegment;
+                throw bi::interprocess_exception("shared memory segment size is 0.");
+            }
+            else
+            {
+                fShmSegment = shmSegment;
+                fSize = curSize;
+                return;
+            }
+        }
+        catch (bi::interprocess_exception&)
+        {
+            if (tries == ShmCreateMaxRetries)
+            {
+                log("BRMManagedShmImpl::BRMManagedShmImpl(): re-creating shared memory segment\
+ b/c of its size == 0. Re-throw.");
+                throw;
+            }
+
+            cerr << "BRMManagedShmImpl::BRMManagedShmImpl(): re-creating shared memory segment\
+ b/c of its size == 0" << endl;
+            usleep(NapTimer);
+        }
+    }
+
+    try
+    {
+        bi::permissions perms;
+        perms.set_unrestricted();
+        fShmSegment = new bi::managed_shared_memory(bi::create_only, keyName.c_str(), fSize,
+            0, // use a default address to map the segment
+            perms);
+        // fSize == 0 on any process startup but managed_shared_memory ctor throws
+        // so control flow doesn't get here.
+        idbassert(fSize > 0);
+    }
+    catch (bi::interprocess_exception &b)
+    {
+        if (b.get_error_code() != bi::already_exists_error)
+        {
+            ostringstream o;
+            o << "BRM caught an exception creating a shared memory segment: " << b.what();
+            log(o.str());
+            throw;
+        }
+        bi::managed_shared_memory* shmSegment = nullptr;
+        try
+        {
+            if (fReadOnly)
+                shmSegment = new bi::managed_shared_memory(bi::open_read_only, keyName.c_str());
+            else
+                shmSegment = new bi::managed_shared_memory(bi::open_only, keyName.c_str());
+        }
+        catch (exception &e)
+        {
+            ostringstream o;
+            o << "BRM caught an exception attaching to a shared memory segment (" << keyName << "): " << b.what();
+            log(o.str());
+            throw;
+        }
+        off_t curSize = shmSegment->get_size();
+
+        idbassert(curSize > 0);
+        idbassert(curSize >= fSize);
+        fShmSegment = shmSegment;
+        fSize = curSize;
+    }
+}
+
+int BRMManagedShmImpl::grow(off_t newSize)
```

Comment:
Overall it's LGTM, I have only one concern about the grow function we implement -  regarding to docs we have to call grow after all other processes are released shared segment (call syscall `umap`), in our case we grow the segment in one process and other processes will know about that segment was grown when they try to grab the `lock` and check segment size, which is kinda breaking the official way of doing this, this potentially could cause an UB, but may be not. It's seems like now it works, but some day kernel people (since boost lib is a just wrapper around linux syscalls) can change something and it will not work anymore, because we use some non documented features, and they are in power to change non documented behavior.